tokio-codec/src/datagram.rs

```diff
@@ -0,0 +1,179 @@
+use crate::{Decoder, Encoder};
+
+use bytes::{BufMut, BytesMut};
+use core::task::{Context, Poll};
+use futures_core::{ready, Stream};
+use futures_sink::Sink;
+use log::trace;
+use std::io;
+use std::pin::Pin;
+
+/// A unified `Stream` and `Sink` interface to an underlying datagram socket, using
+/// the `Encoder` and `Decoder` traits to encode and decode frames.
+///
+/// Even for datagram-based sockets, higher-level code usually wants to
+/// batch these into meaningful chunks, called "frames". This method layers
+/// framing on top of this socket by using the `Encoder` and `Decoder` traits to
+/// handle encoding and decoding of messages frames. Note that the incoming and
+/// outgoing frame types may be distinct.
+///
+/// This function returns a *single* object that is both `Stream` and `Sink`;
+/// grouping this into a single object is often useful for layering things which
+/// require both read and write access to the underlying object.
+#[must_use = "sinks do nothing unless polled"]
+#[derive(Debug)]
+pub struct DatagramFramed<C, S, A> {
+    socket: S,
+    codec: C,
+    rd: BytesMut,
+    wr: BytesMut,
+    out_addr: Option<A>,
+    flushed: bool,
+}
+
+pub trait DatagramSocket<A> {
```

Comment:
I believe `A` should be an associated type. We could call this `Address`?

I also don't think `tokio-net` should have a dependency on `tokio-codec`, so maybe lets move this trait into `tokio-io` and name it `AsyncDatagram` (matching `AsyncRead` / `AsyncWrite`).

Reply:
You're right, that should absolutely be an associated type :) I moved `AsyncDatagram` to `tokio-io`.